icu4c/source/common/umutex.cpp

```diff
@@ -43,8 +43,8 @@ U_NAMESPACE_BEGIN
 
 // The ICU global mutex. Used when ICU implementation code passes NULL for the mutex pointer.
 static UMutex *globalMutex() {
```

Comment:
To facilitate future cleanup efforts, please move the definition of this and all other mutexes using the same mechanism to headers.

Reply:
To avoid the ordering issue, you'll probably need to pull all the mutex declarations in one place so that you can then destroy them in a predictable manner on unload. But yes that can wait until that problem is addressed.

---
In reply to: [279954961](https://github.com/unicode-org/icu/pull/620#discussion_r279954961) [](ancestors = 279954961)

Reply:
> To avoid the ordering issue, you'll probably need to pull all the mutex declarations in one place ...

ICU already has a clean-up mechanism. It does not work like you are thinking.

Each ICU module/service/file/whatever that needs cleanup registers a cleanup function with ICU's clean up service. The registration includes an ordering, to avoid dependency problems.

Exactly what each module's cleanup function needs to handle is not surfaced globally; that is purely the internal business of that module. Exporting that kind of internal detail to more global scopes would not be desirable.

Anything unloading the ICU DLLs already must use u_cleanup(), to avoid memory leaks from ICU's internal caches. Many ICU modules using mutexes already have cleanup functions in place already that could be extended.

Note also, ICU must continue to operate correctly if use resumes after calling u_cleanup() (without unloading and reloading the library). This means that function local static init cannot be used as a lazy init mechanism for anything that is deleted by u_cleanup().

In any event, this PR is not the right place to design mutex cleanup. It's just to get us back to ICU 63 functionality, which is currently broken and causing immediate problems.